### ISSUE TITLE
Wrong error message for nested expectations

### ISSUE THREAD

**Issue by severinraez** [brigade/db-query-matchers#11]
Hi,

I spent some time debugging weird failing expectations in our tests. The message was 

    Expected 4 queries, but 4 were made

Turns out if you expect something in the block matched by *make_database_queries*, the expectation will be described by the *failure_message* block in *make_database_queries.rb* - which leads to this puzzling message.

To illustrate, I added an example failing with 

    expected 1 query, but 1 was made:
           SELECT  "cats".* FROM "cats"  ORDER BY "cats"."id" ASC LIMIT 1

in my fork at https://github.com/severinraez/db-query-matchers

I'm afraid I don't see how to fix this. Could you point me in the right direction? I'll happily submit a patch.

EDIT: Commit containing the spec: https://github.com/severinraez/db-query-matchers/commit/dadbe068270f30aedfb488275965a61d78966068

**Comment by trotzig** [comment]
Interesting! Thanks for narrowing it down. 

I'm not completely sure, but this could be a problem with any rspec matcher that `supports_block_expectations`. I've managed to reproduce it with a custom matcher in this bit of rspec:

```ruby

RSpec::Matchers.define :foo do
  match do |actual|
    actual.call
    true
  end

  def supports_block_expectations?
    true # or some logic
  end

  def failure_message
    'I failed'
  end
end

describe '#reraise' do
  context 'when some other expectation in the block fails' do
    subject {
      raise RSpec::Expectations::ExpectationNotMetError.new('other')
    }

    it 'reraises the error' do
      expect do
        expect { subject }.to foo
      end.to raise_error(RSpec::Expectations::ExpectationNotMetError, /other/)
    end
  end
end
```

This will fail with the following message:
```
F

Failures:

  1) #reraise when some other expectation in the block fails reraises the error
     Failure/Error:
       expect do
         expect { subject }.to foo
       end.to raise_error(RSpec::Expectations::ExpectationNotMetError, /other/)

       expected RSpec::Expectations::ExpectationNotMetError with message matching /other/, got #<RSpec::Expectations::Expecta
tionNotMetError: I failed> with backtrace:
         # ./spec/db_query_matchers/make_database_queries_spec.rb:26:in `block (4 levels) in <top (required)>'
         # ./spec/db_query_matchers/make_database_queries_spec.rb:25:in `block (3 levels) in <top (required)>'
     # ./spec/db_query_matchers/make_database_queries_spec.rb:25:in `block (3 levels) in <top (required)>'

Finished in 0.01832 seconds (files took 1.15 seconds to load)
1 example, 1 failure
```
There might be a way to get exceptions to bubble through matchers, but at the moment I don't know how.

**Comment by trotzig** [comment]
Fixed by 0f42325ca45f1
